### ISSUE TITLE
Assertion in test_kdtree failed

### ISSUE THREAD

**Issue by bialasjaroslaw** [thomgrand/tf_kdtree#1]
After small modifications regarding linking to `Eigen` and `Tensorflow` I was able to compile code and run tests. However there is a lot of lines like one below that are suggesting that something is not right.

> Generated points...
/home/jbialas/Projects/tf_kdtree/src/kdtree.hpp:310: void TreeTraversal<T, dims>::moveToLeftChild() [with T = double; unsigned char dims = (unsigned char)'']: block: [20,4,0], thread: [0,0,0] Assertion current_level < partition_info->levels failed.

I am using GCC-11.3.0 with CUDA 11.8.
I am linking with system Eigen 3.4.0, Assimp 5.2.5 and Tensorflow 2.10.0

I will be happy to provide more information and even provide some contribution if that will help, but maybe this is just a simple fix.

**Comment by thomgrand** [comment]
Thanks for reporting this problem. I just retried building the library from the ground up and also noticed that it's not compatible with the current master branch of Eigen. I'll fix the library to Eigen 3.4 to avoid these errors in the future.
I can not however reproduce the error you mentioned, also in Debug compilation. Do you also have trouble with the CPU implementation or did you modify any other parts? It seems the assertion only fails in CUDA.
Here's my configuration:
- GCC 9.4
- CUDA 11.2
- (Tensorflow 2.10)

test_kdtree actually does not use python iirc and is purely for testing the C++ implementation.

**Comment by bialasjaroslaw** [comment]
Thank you for fast response.
I checked it, and there is nothing wrong reported from application if I remove lines with GPU test from `test_kdtree.cpp` - if that is what you mean. However, it is enough to leave line with `KDTreeKNNGPUSearch` to see 435 lines with error from previous message (only block number differs)
I will check how it works with older GCC and CUDA to determine if this is related to compiler or CUDA version.

**Comment by bialasjaroslaw** [comment]
I was able to test following setups
CUDA 11.2 + GCC9 - OK
CUDA 11.3 + GCC9 - TEST FAIL
CUDA 11.4 + GCC9 - TEST FAIL
CUDA 11.8 + GCC11 - TEST FAIL

I don't have working GCC10 right now, but I think this is not a GCC's fault. I have analyzed logs from compilation for first two cases and there is a lot warnings for the 11.2 while less with 11.3. Lot of them seem to be harmless
```bash
/usr/include/tensorflow/Eigen/src/Core/Reshaped.h(139): warning: 
__host__ annotation is ignored on a function("ReshapedImpl") 
that is explicitly defaulted on its first declaration
```
However there are warnings that are more serious and they are still warnings
```bash
/home/jbialas/Projects/tf_kdtree/src/kdtree.hpp(313): warning: 
calling a __host__ function("T1 TreeTraversal<float, (unsigned char)1> 
::compLeftLeafInd<unsigned int, unsigned int> (T1, T2)") 
from a __host__ __device__ function("TreeTraversal<float, (unsigned char)1> 
::getLeftLeaf const") is not allowed
```

It is even more interesting that this specific warning is not present for 11.2 build but appears in build log from 11.3. 
On the other hand there are few warnings like one below that suggest something bad might happen
```bash
warning: dynamic initialization is not supported for a function-scope 
static __shared__ variable within a __device__/__global__ function
```
And again I don't know if for some reason from any version of CUDA this might cause some issues, but there is nothing more that can indicate problems with the code.

Any hints what could be done to diagnose cause of a problem in case of CUDA 11.3 and newer?

**Comment by thomgrand** [comment]
Thanks for the thorough analysis. I can confirm that I run into the same problem with CUDA 11.4. I found a few hints online that [higher versions of CUDA break the compatibility with Eigen](https://forums.developer.nvidia.com/t/breaking-change-for-eigen-code-in-nvcc-11-3-and-later-vs-nvcc-11-2-and-earlier/186800), but no pointers what causes the problem or how to fix it.
Since Eigen is only used for some functions in the CUDA kernels (subtraction, and squared norm mostly), I think it's worth considering a simple replacement.

**Comment by bialasjaroslaw** [comment]
I found a solution, and it is not related to Eigen. 
I have analyzed warnings provided by nvcc and there were a lot of them mentioning that some functions can not be called from device context. This was odd as there were nothing illegal with them (in my opinion). I noticed that even though I have add `--expt-relaxed-constexpr` flag to `CUDA_NVCC_FLAGS`, couple templated functions were still reported as problematic. Turns out it was enough to mark them explicitly CUDA_HOSTDEV`. 
I will provide a fix as a PR if you don't mind.
